app/org/sagebionetworks/bridge/services/StudyService.java

```diff
@@ -670,4 +732,120 @@ protected EmailTemplate sanitizeEmailTemplate(EmailTemplate template) {
+        if (data == null) {
+            throw new BadRequestException("Email verification token has expired (or already been used).");
+        }
+
+        // Make sure the study's current consent notification email matches the email saved in the verification data.
+        // If the study's consent notification email is updated, the caller might still be using an older verification
+        // email.
+        Study study = getStudy(data.getStudyId());
+        if (!study.getConsentNotificationEmail().equals(data.getEmail())) {
+            throw new BadRequestException("Email verification token does not match consent notification email.");
+        }
+
+        // Set verified flag and update study.
+        study.setConsentNotificationEmailVerified(true);
+        updateAndCacheStudy(study);
+    }
+
+    // Creates a random token for consent notification email verification. Package-scoped so it can be mocked by unit
+    // tests.
+    String createTimeLimitedToken() {
+        return SecureTokenGenerator.INSTANCE.nextToken();
+    }
+
+    // Helper method to save consent notification email verification data from the cache.
+    private void saveVerification(String sptoken, VerificationData data) {
+        checkArgument(isNotBlank(sptoken));
+        checkNotNull(data);
+
+        try {
+            cacheProvider.setObject(sptoken, BridgeObjectMapper.get().writeValueAsString(data),
+                    CONSENT_NOTIFICATION_VERIFICATION_EXPIRE_IN_SECONDS);
+        } catch (IOException e) {
+            throw new BridgeServiceException(e);
+        }
+    }
+
+    // Helper method to fetch consent notification email verification data from the cache.
+    private VerificationData restoreVerification(String sptoken) {
+        checkArgument(isNotBlank(sptoken));
+
+        String json = cacheProvider.getObject(sptoken, String.class);
+        if (json != null) {
+            try {
+                cacheProvider.removeObject(sptoken);
+                return BridgeObjectMapper.get().readValue(json, VerificationData.class);
+            } catch (IOException e) {
+                throw new BridgeServiceException(e);
+            }
+        }
+        return null;
+    }
+
+    // Verification data for consent notification email.
+    private static class VerificationData {
+        private final String studyId;
+        private final String email;
+
+        @JsonCreator
+        VerificationData(@JsonProperty("studyId") String studyId, @JsonProperty("email") String email) {
+            checkArgument(isNotBlank(studyId));
+            checkArgument(isNotBlank(email));
+            this.studyId = studyId;
+            this.email = email;
+        }
+
+        // Study ID that we want to verify email for.
+        public String getStudyId() {
+            return studyId;
+        }
+
+        // Email address that we want to verify.
+        public String getEmail() {
+            return email;
+        }
+    }
```

Comment:
ditto, already in AccountWorkflowService. Let's use what's there

Reply:
The one in AccountWorkflowService uses userId rather than email. The I could use email in place of userId, but that seems more like a hack.